components/scream/src/physics/p3/atmosphere_microphysics.hpp

```diff
@@ -243,6 +243,13 @@ class P3Microphysics : public AtmosphereProcess
       } // for ipack
       precip_liq_surf_mass(icol) += precip_liq_surf_flux(icol) * PC::RHO_H2O * m_dt;
       precip_ice_surf_mass(icol) += precip_ice_surf_flux(icol) * PC::RHO_H2O * m_dt;
+
+      // Set appropriate boundary fluxes for energy and mass conservation checks.
+      // Any boundary fluxes not included in P3 interface are set to 0.
```

Comment:
Might be good to provide a comment here explaining why these fluxes are not accumulated like the ones above.

Reply:
ahh Technically they should be accumulated, I think. And then reset to 0 after P3 is run (like `precip_liq/ice_surf_mass`). Having P3 internal subcycle be 1 hides this issue. I'll update.

Reply:
Wait, I'm actually not sure now. In `micro_p3_interface` these values appear to be set (and not accumulated) each time `micro_p3_tend` is run, even during subcycling.

Reply:
Right, it's fine not to accumulate these; I just meant you should provide a comment explaining why, since right above these lines, fluxes are accumulated.

Reply:
Ok, I'll add a comment.

Reply:
I'm actually not sure why this is the case (outside of "it's what eam does"). I'm guessing it has to with the fact that we are only checking the mass/energy change during a single step, so we don't need to accumulate from the previous steps.

Reply:
Yes, exactly: In both EAM and EAMxx, the macro- and microphysics routines can be subcycled together, so their fluxes have to be accumulated over the physics time step. But the energy and mass checking is done at each call to each physics process, so no accumulation is done. This is what you should document for future readers.

Reply:
What if we have internal p3-only subcycles? In that case the checker is run after multiple p3 calls.
```
In input.yaml:

  p3:
     num_subcycles: n

In atmosphere_process.cpp:run(dt):

  compute_column_conservation_check_data()
  for (num_subcycles) {
     run_impl(dt/num_subcycles)
  }
  run_column_conservation_check()
```
We would need to accumulate in the case, right? Either that or move the conservation stuff inside the subcycle loop:
```
  for (num_subcycles) {
     compute_column_conservation_check_data()
     run_impl(dt/num_subcycles)
     run_column_conservation_check()
  }
```

Reply:
> move the conservation stuff inside the subcycle loop

Yes.

But there is perhaps more complexity in the AD than is really necessary. There isn't a need to support process-level subcycling, just grouped-process subcycling. (Process-level subcycling is better handled in the process.) Considering the hierarchy of AtmosphereProcess(Group)::run(_impl) calls, have you thought about removing conservation-check calls from the AtmosphereProcess run routine and instead making it something that a client process would call in their run_impl routine? Doing so would have the advantages of (1) making clear in the code the valid time extent of the mass and energy fluxes and (2) avoiding conservation-related calls in nodes of the AP(G) tree that clearly don't logically support the check. @bartgol, thoughts?

Reply:
Right now I require that a process calling the check logically should call the check. I.e., is a physics process that defines boundary fluxes. So if the user tries to enable the conservation check on a group, or dynamics, or a physics process that doesn't define boundary fluxes, it throws an error. 

I think your suggestion would require a lot of code duplication, since each individual process would need to do the same process of asking "do I have the check? If so, precompute data" and then "do I have the check? If so, run the check".

Another option in a completely different direction would be to allow the user to set `enable_column_conservation_checks" at the physics group level (with a tolerance), and then atmosphere_process_group decides which individual process gets the check. A downside here is that, the a process may want to run the check, but the creator of the process doesn't realize that they needed to define boundary fluxes and is never told.

Reply:
Ok, I'm convinced. The critical driver here is to make sure a conservation check is done for a single process call and nothing else or, alternatively, to accumulate fluxes for the conservation check at the right level of granularity. So your suggestion of moving the calls into the subcycling loop is probably best.

Reply:
It is true that, if a single individual process had to be subcycled, then it can be done inside run_impl. That's precisely what HommeDynamics currently does, with `nsplit` (although homme has a pre-subcycling phase (`homme_pre_process`) and a post-subcycling phase (`homme_post_process`) that would not otherwise be performed had it used subcycling from the base class, so this is a bit different from a pure subcycling). OTOH, keeping it at the AD level seemed like an easy enough thing to do, which costed nothing to do. I am also thinking that, if/when we support parallel splitting for processes, we _might_ need new atm proc classes to wrap a "remote" process. I don't have a design in mind, but keeping things general might prove useful down the road.

Reply:
But I don't understand @tcclevenger point. Why can't we do what andrew suggested? I don't understand `each individual process would need to do the same process of asking "do I have the check? If so, precompute data"`. If a developer for parametrization FOO _knows_ that FOO should conserve mass/energy, they simply call the setup/check method at the top/bottom of run_impl.

Reply:
My only point is that, the code in the individual processes interface will all be identical copies of each other (and to what is currently in `atmosphere_process.cpp::run()`).  Whereas right now the developer just adds the input.yaml line `enable_column_conservation_check: true` if they want a certain process checked for conservation, and the rest is handled internally.

We definitely can do what Andrew suggested, but I think it would result in more lines of duplicated code between the processes.